be/src/vec/common/sort/sorter.cpp

```diff
@@ -0,0 +1,252 @@
+    }
+    if (_reach_limit()) {
+        RETURN_IF_ERROR(_do_sort());
+    }
+    return Status::OK();
+}
+
+Status FullSorter::prepare_for_read() {
+    if (_state->unsorted_block->rows() > 0) {
+        RETURN_IF_ERROR(_do_sort());
+    }
+    _state->build_merge_tree(_sort_description);
+    return Status::OK();
+}
+
+Status FullSorter::get_next(RuntimeState* state, Block* block, bool* eos) {
+    if (_state->sorted_blocks.empty()) {
+        *eos = true;
+    } else if (_state->sorted_blocks.size() == 1) {
+        if (_offset != 0) {
+            _state->sorted_blocks[0].skip_num_rows(_offset);
+        }
+        block->swap(_state->sorted_blocks[0]);
+        *eos = true;
+    } else {
+        RETURN_IF_ERROR(_state->merge_sort_read(state, block, eos));
+    }
+    return Status::OK();
+}
+
+Status FullSorter::_do_sort() {
+    Block block = _state->unsorted_block->to_block(0);
+    RETURN_IF_ERROR(partial_sort(block));
+    // dispose TOP-N logic
+    if (_limit != -1) {
+        // Here is a little opt to reduce the mem uasge, we build a max heap
+        // to order the block in _block_priority_queue.
+        // if one block totally greater the heap top of _block_priority_queue
+        // we can throw the block data directly.
+        if (_state->num_rows < _limit) {
+            _state->sorted_blocks.emplace_back(std::move(block));
+            _state->num_rows += block.rows();
+            _block_priority_queue.emplace(_pool->add(
+                    new SortCursorImpl(_state->sorted_blocks.back(), _sort_description)));
+        } else {
+            SortBlockCursor block_cursor(_pool->add(new SortCursorImpl(block, _sort_description)));
+            if (!block_cursor.totally_greater(_block_priority_queue.top())) {
+                _state->sorted_blocks.emplace_back(std::move(block));
+                _block_priority_queue.push(block_cursor);
+            }
+        }
+    } else {
+        // dispose normal sort logic
+        _state->sorted_blocks.emplace_back(std::move(block));
+    }
+    _state->reset_block();
+    return Status::OK();
+}
+
+TopNSorter::TopNSorter(SortDescription& sort_description, VSortExecExprs& vsort_exec_exprs,
+                       int limit, int64_t offset, ObjectPool* pool, std::vector<bool>& is_asc_order,
+                       std::vector<bool>& nulls_first, const RowDescriptor& row_desc)
+        : Sorter(sort_description, vsort_exec_exprs, limit, offset, pool, is_asc_order,
+                 nulls_first),
+          _state(std::unique_ptr<MergeSorterState>(new MergeSorterState(row_desc, offset))) {}
+
+Status TopNSorter::append_block(Block* block, bool* mem_reuse) {
+    DCHECK(block->rows() > 0);
+    RETURN_IF_ERROR(_do_sort(block, mem_reuse));
+    return Status::OK();
+}
+
+Status TopNSorter::prepare_for_read() {
+    _state->build_merge_tree(_sort_description);
+    return Status::OK();
+}
+
+Status TopNSorter::get_next(RuntimeState* state, Block* block, bool* eos) {
```

Comment:
seems the code is same as `FullSorter` `get_next` may should not virtual function and in `Sorter `